### PR TITLE
LIVE-1662: investigate adding Storyshots snapshot testing

### PR DESCRIPTION
**Pull request by joecowton1**
## Why are you doing this?

Investigating the addition of snapshot testing using the Storybook addon "Storyshots" and Jest.

For anyone who hasn't come across this before there's more info here:
https://www.npmjs.com/package/@storybook/addon-storyshots

The setup is very simple, we've had to update the Storybook config files to use their new conventions but otherwise all this requires is a `storyshots.test.ts` file that initialises the tests .

Each time we run the test script now Jest will create a snapshot for every file that has a `*.stories.tsx` file - giving us a way to monitor for changes in the UI.

Creating the snapshots adds around 2.5 seconds to the time the test script takes to run and currently generates 65 snapshots, all contained within one file.

If a snapshot has been changed Jest will flag it as below:

We can either fix this by updating the snapshot (`npm run test -- -u`) or fix the code so the snapshot doesn't change.

**TL;DR** Storyshot testing adds a snapshot for every component covered by a story. The test script takes a couple of seconds longer to run, but we gain a lot of UI test coverage in return for adding minimal config to the project.

## Screenshots

| Before | After |
| --- | --- |
|  |  |

### CODE REVIEW

**Comment by JamieB-gu** [comment]
FYI @oliverlloyd @gtrufitt @sndrs @SiAdcock

**Comment by oliverlloyd** [comment]
That's interesting, I've not seen this before. I'd be keen to hear you feedback on how this goes.

I've had bad experiences with pure jest snapshots before where instead of writing classic unit tests everything was a snapshot becasue that was easier but then that just resulted in everyone pressing `u` a lot and not having the time (or being bothered enough) to review all the html changes 😞 

But I wonder if using it in the context of Storybook solves these issues? That would be ace.

**Comment by joecowton1** [comment]
Hi @oliverlloyd - yeah I've certainly experience similar scenarios to the one you mentioned so in some ways I share your scepticism.

At the same time the snapshots we're generating here are pretty clean and descriptive - in my experience noisy snapshots just aren't worth having - that we're also creating the snapshots based on specific use cases (by format / template) should be helpful in terms of flagging unintended consequences when making amends.

Aware a lot of what I'm saying verges on the optimistic end of things, but I think it's at least worth trying as a trial. Worst case we remove the snapshot initialiser and nothing is lost.

@JamieB-gu do we have your blessing to give it a go?

**Comment by oliverlloyd** [comment]
> Aware a lot of what I'm saying verges on the optimistic end of things, but I think it's at least worth trying as a trial. Worst case we remove the snapshot initialiser and nothing is lost.

This is a really good point 👍 

I also fully agree that tight, focussed small snapshots are the way to go and it's the larger rambling ones that cause the pain so this context alignment to the stories should guard against that.

Please do give it a try and let us know how it went

**Review by JamieB-gu — APPROVED** [review]
Agree with what's been said above 🚀 

Couple questions:

- Do you see this as another tool for snapshotting that sits alongside the Chromatic tests?
- Do you know if we can integrate this with Emotion to get styles as part of the snapshots? Here's [an example](https://github.com/guardian/image-rendering/blob/e9f1633c7f36409b1a2c8d591c53ac3adff4d1e1/package.json#L78) from image-rendering, and the [resulting snapshots](https://github.com/guardian/image-rendering/blob/e9f1633c7f36409b1a2c8d591c53ac3adff4d1e1/src/__snapshots__/sizes.test.tsx.snap)

**Comment by joecowton1** [comment]
> * Do you see this as another tool for snapshotting that sits alongside the Chromatic tests?

@JamieB-gu was speaking to @faresite about this earlier - with the Emotion integration you mentioned in place we definitely think this will be useful in spotting UI regressions. Agree there's some crossover with Chromatic but this should be a lot more comprehensive and detailed than what we're currently using Chromatic for. Does that this makes sense?

File: .storybook/main.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+module.exports = {
+    stories: ['../src/**/*.stories.@(js|mdx|ts|tsx)'],
```
[the remark this answers is not in the document]
Comment (joecowton1) [reply .storybook/main.js]: Good shout, fixed.

**Comment by sndrs** [comment]
this is really interesting – i feel an CSI presentation coming on...

**Review by faresite — APPROVED** [review]
